src/test/java/org/opensearch/security/identity/ContextProvidingPluginSubjectTests.java

```diff
@@ -0,0 +1,103 @@
+import org.opensearch.plugins.Plugin;
+import org.opensearch.security.user.PluginUser;
+import org.opensearch.threadpool.TestThreadPool;
+import org.opensearch.threadpool.ThreadPool;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.opensearch.security.support.ConfigConstants.OPENDISTRO_SECURITY_USER;
+import static org.junit.Assert.assertNull;
+
+public class ContextProvidingPluginSubjectTests {
+    static class TestIdentityAwarePlugin extends Plugin implements IdentityAwarePlugin {
+
+    }
+
+    @Test
+    public void testSecurityUserSubjectRunAs() throws Exception {
+        final ThreadPool threadPool = new TestThreadPool(getClass().getName());
+
+        final Plugin testPlugin = new TestIdentityAwarePlugin();
+
+        final PluginUser pluginUser = new PluginUser(testPlugin.getClass().getCanonicalName());
+
+        ContextProvidingPluginSubject subject = new ContextProvidingPluginSubject(threadPool, Settings.EMPTY, testPlugin);
+
+        assertThat(subject.getPrincipal().getName(), equalTo(testPlugin.getClass().getCanonicalName()));
+
+        assertNull(threadPool.getThreadContext().getTransient(OPENDISTRO_SECURITY_USER));
+
+        subject.runAs(() -> {
+            assertThat(threadPool.getThreadContext().getTransient(OPENDISTRO_SECURITY_USER), equalTo(pluginUser));
+            return null;
+        });
+
+        assertNull(threadPool.getThreadContext().getTransient(OPENDISTRO_SECURITY_USER));
+
+        SecurityUserSubjectTests.terminate(threadPool);
+    }
+
+    @Test
+    public void testPluginContextSwitcherRunAs() throws Exception {
+        final ThreadPool threadPool = new TestThreadPool(getClass().getName());
+
+        final Plugin testPlugin = new TestIdentityAwarePlugin();
+
+        final PluginContextSwitcher contextSwitcher = new PluginContextSwitcher();
+
+        final PluginUser pluginUser = new PluginUser(testPlugin.getClass().getCanonicalName());
+
+        ContextProvidingPluginSubject subject = new ContextProvidingPluginSubject(threadPool, Settings.EMPTY, testPlugin);
+
+        contextSwitcher.initialize(subject);
+
+        assertNull(threadPool.getThreadContext().getTransient(OPENDISTRO_SECURITY_USER));
+
+        subject.runAs(() -> {
+            assertThat(threadPool.getThreadContext().getTransient(OPENDISTRO_SECURITY_USER), equalTo(pluginUser));
+            return null;
+        });
+
+        assertNull(threadPool.getThreadContext().getTransient(OPENDISTRO_SECURITY_USER));
+
+        SecurityUserSubjectTests.terminate(threadPool);
+    }
+
+    @Test
+    public void testPluginContextSwitcherUninitializedRunAs() throws Exception {
+        final ThreadPool threadPool = new TestThreadPool(getClass().getName());
+
+        final PluginContextSwitcher contextSwitcher = new PluginContextSwitcher();
+
+        threadPool.getThreadContext().putHeader("foo", "bar");
+
+        // should be no noop since contextSwitch is uninitialized
```

Comment:
would the expected value still be "bar" for "foo" if contextSwitcher was initialized?

Reply:
If contextSwitcher is initialized then it should expect all TC headers to be nullified except for the transient OPENDISTRO_SECURITY_USER header which is populated with the pluginUser.